### ISSUE TITLE
p4nzenc256v32 sometimes produces different results

### ISSUE THREAD

**Issue by pps83** [powturbo/TurboPFor#60]
While making some changes in my code, I tried to verify that the resulting binary data (compressed with p4nzenc256v32) is identical. To my surprise, the data had many changes as if something corrupted it. I tried to dump hex of compressed data and it clearly had a few modified bytes here and there. What's even more surprising, when I uncompressed the data using p4nzdec256v32 output from these two different buffers was identical.
Any idea why, is that normal? If so, what makes it produce different outputs with identical inputs?

**Comment by powturbo** [comment]
It's possible that some bits in the last byte of a block contain random bits. 
These bits are not needed in decoding. 
Ex. you can pack 33 integers of 1 bit in 5 bytes and byte 5 has 7 unused bits.

**Comment by pps83** [comment]
In my case I have differences in the middle of entire 3KB compressed block. Can trailing bits cause differences within compressed block?

My 32bit data input has multiple ints, no trailing random bits.

**Comment by pps83** [comment]
I tried to add 64 int=0 values at the end of input buffer to make sure that if p4nzenc256v32 over-reads it would always have the same input, but I still get some compressed buffer with random bits. Also, it seems that these random bits are always in the same location

**Comment by pps83** [comment]
I'm fairly sure that optimized build of p4nzenc256v32 has some bugs that results in random output.
I added a test program that shows that: https://github.com/pps83/TurboPFor/blob/p4nzenc256v32-bug/vs/p4nzenc256v32-bug.cpp (you can checkout `p4nzenc256v32-bug` branch: https://github.com/pps83/TurboPFor/tree/p4nzenc256v32-bug and build/run `p4nzenc256v32-bug` VS project to see it).

In short, I have a sample array that I compress 10 times and then compare produced outputs. In debug x64 and x86 output is always identical (hash: `228720909521576241`), but in release all the outputs are different. Interestingly, in x86 only first output is different, all others are identical. In x64 most of the outputs are different:
```
 out8[0] hash:14659539012829149947
 out8[1] hash:16632647345969194266
 out8[2] hash:5095059712619303890
 out8[3] hash:12388682344188382616
 out8[4] hash:14659539012829149947
 out8[5] hash:16632647345969194266
 out8[6] hash:5095059712619303890
 out8[7] hash:12388682344188382616
 out8[8] hash:14659539012829149947
 out8[9] hash:16632647345969194266
```

The code somewhere ends up using uninitialized avx/sse registers which results in some feedback between calls to p4nzenc256v32. If I add `_mm256_zeroall();` at the beginning of `p4nzenc256v32` function, then output in x64 proves my theory, and feedback is reduced and looks like that:

```
 out8[0] hash:14659539012829149947
 out8[1] hash:5799741302505786334
 out8[2] hash:5799741302505786334
 out8[3] hash:5799741302505786334
 out8[4] hash:5799741302505786334
 out8[5] hash:5799741302505786334
 out8[6] hash:5799741302505786334
 out8[7] hash:5799741302505786334
 out8[8] hash:5799741302505786334
 out8[9] hash:5799741302505786334
```

In my actual code I can use FastPFor instead, and I see these random output changes only if I use TurboPFor

**Comment by powturbo** [comment]
Well, this is not really a bug and I'm aware since a long time that the output is not deterministic because of some unused random bits. TurboPFor is decoding correctly the encoded input.
I'll take a look at it the next time updating TurboPFor.

**Comment by pps83** [comment]
Yes, it decodes with no issues. What unused random bits are you referring to? Did you mean used random bits? In my case, I would totally prefer to have deterministic encoder, I'd prefer to compress with Debug build or anything even if it was 50% slower. In my case, after performing complex manipulations I verify correctness by diffing outputs and my outputs are always different since I started using turbopfor
Is it easy to clean those unused undeterministic bits?

**Comment by magenbluten** [comment]
+1 for having deterministic output. if it is not possible to achieve then please document this somewhere. ( i was planning to use turbopfor for my upcoming project ).

**Comment by powturbo** [comment]
As already stated, I'll take a look at it the next time updating TurboPFor.

**Comment by powturbo** [comment]
I've made a change now in "vp4c.c" and tested this with your test application and it seems working now.
Thank you for reporting and your contribution.

**Comment by pps83** [comment]
FYI, this change again introduced trailing spaces and tabs:
https://github.com/powturbo/TurboPFor/commit/363fcc03283c67e1428ebc9164fe65be377e9711

This one looks like shouldn't have been committed:
https://github.com/powturbo/TurboPFor/commit/1ae28cc6c45b609841adc29a63b9fc236d2bf53b

**Comment by pps83** [comment]
Fixed by 086419ab05fb0820899515a3b34d4bbdeeab1618

**Comment by powturbo** [comment]
Thank you very much for your help

**Comment by pps83** [comment]
That one was ok, but when I tried to verify the results over large amount of data I still got the same amount of differences in other places.
This is sample program: https://github.com/pps83/TurboPFor/blob/p4nzenc256v32-bug-2/vs/p4nzenc256v32-bug.cpp

Note, that it uses `p4nzenc32`, also, it doesn't easily reproduce for some reason (all the buffers are identical), however, Release of x64 is different from Debug-x64/Debug-x86/Release-x86 all of which produce identical outputs. If I place `_mm256_zeroall();` at the beginning of `p4nzenc32`, then x64 starts to produce identical results.

**Comment by pps83** [comment]
The test program is in `p4nzenc256v32-bug-2` branch:
https://github.com/pps83/TurboPFor/tree/p4nzenc256v32-bug-2

**Comment by powturbo** [comment]
I've tested your bug-2 program and I can't reproduce the errors. I've initialized all the arrays used in vp4c.c encoding. Please, clean, rebuild all debug & release versions and try again.

**Comment by pps83** [comment]
I'm sure about my build.
In that updated `p4nzenc256v32-bug.cpp` I get this hash for Debug-x64, Debug-x86, Release-x86: `8447788239561535011`
However, I get `10824744585826795417` for Release-x64. It's difficult to repro different result bug, but in my real app I have multiple threads doing the same thing and the problem happen easily (multiple threads => different stack memory). Also, note, if you add `_mm256_zeroall();` at the beginning of `p4nzenc32`:
```
size_t TEMPLATE2(P4NENC, USIZE)(uint_t *__restrict in, size_t n, unsigned char *__restrict out) {
  _mm256_zeroall(); // <<=== added
  unsigned char *op = out;
  uint_t *ip, start = *in++;
  if(!n)
```

then Release-x64 build starts to produce the same result as other builds. Note, that in my real app this fix doesn't fix the problem. Also, note, I see the problem only with `p4nzenc32`, and I don't see it with `p4nzenc256v32`

**Comment by pps83** [comment]
Your update with extra buffer inits fixes this array, but doesn't fix other problems I'll update the data shortly

**Comment by pps83** [comment]
Try latest https://github.com/pps83/TurboPFor/commits/p4nzenc256v32-bug-2, I changed it to use threads, fails easily now (also I merged all your updates)

**Comment by pps83** [comment]
this is what I'm getting now:

```out8 size:1333

 out8[ 0] hash:14212870176462373963
 out8[ 2] hash:12781980896290985357
 out8[ 7] hash:2058967258774365968
 out8[12] hash:14773275750134418099
 out8[27] hash:12694884597877758148

!! HAS DATA MISMATCH !!
```

**Comment by pps83** [comment]
Also, my tests show that only `p4nzenc32` has the problem.
I tried these functions: 
p4nzenc256v32, p4nzenc128v32, p4nzenc32
p4nenc256v32, p4nenc128v32, p4nenc32
p4ndenc256v32, p4ndenc128v32, p4ndenc32
p4nd1enc256v32, p4nd1enc128v32, p4nd1enc32

**Comment by powturbo** [comment]
I'm working on linux and with your latest test program with threads, it doesn't show any mismatches.
These random unused bits can only occurs in the scalar functions when the number of the encoded integers is not  multiple of 32 and not in the SIMD functions (*128venc and *256venc).
I've made now a change in the delta (d,z) functions. If it's still not working then I must write a dedicated test to check the unused bits.

**Comment by pps83** [comment]
This commit fixed the problem: fac4c732f6cc6c0a9c5a5cb4333fc9ff5297fb4d

**Comment by alexb5dh** [comment]
@powturbo , any chance fix can be applied again?
